pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AvgPrecisionAggregationFunction.java

```diff
@@ -0,0 +1,416 @@
+          AvgPrecisionPair innerPair = acum == null ? new AvgPrecisionPair() : acum;
+          for (int i = from; i < to; i++) {
+            innerPair.apply(BigDecimal.valueOf(intValues[i]));
+          }
+          return innerPair;
+        });
+
+        break;
+      case LONG:
+        long[] longValues = blockValSet.getLongValuesSV();
+
+        avgPair = foldNotNull(length, blockValSet, null, (acum, from, to) -> {
+          AvgPrecisionPair innerPair = acum == null ? new AvgPrecisionPair() : acum;
+          for (int i = from; i < to; i++) {
+            innerPair.apply(BigDecimal.valueOf(longValues[i]));
+          }
+          return innerPair;
+        });
+
+        break;
+      case FLOAT:
+      case DOUBLE:
+      case STRING:
+        String[] stringValues = blockValSet.getStringValuesSV();
+
+        avgPair = foldNotNull(length, blockValSet, null, (acum, from, to) -> {
+          AvgPrecisionPair innerPair = acum == null ? new AvgPrecisionPair() : acum;
+          for (int i = from; i < to; i++) {
+            innerPair.apply(new BigDecimal(stringValues[i]));
+          }
+          return innerPair;
+        });
+
+        break;
+      case BIG_DECIMAL:
+        BigDecimal[] bigDecimalValues = blockValSet.getBigDecimalValuesSV();
+
+        avgPair = foldNotNull(length, blockValSet, null, (acum, from, to) -> {
+          AvgPrecisionPair innerPair = acum == null ? new AvgPrecisionPair() : acum;
+          for (int i = from; i < to; i++) {
+            innerPair.apply(bigDecimalValues[i]);
+          }
+          return innerPair;
+        });
+
+        break;
+      case BYTES:
+        // Serialized AvgPrecisionPair
+        byte[][] bytesValues = blockValSet.getBytesValuesSV();
+        avgPair = new AvgPrecisionPair();
+        for (int i = 0; i < length; i++) {
+          AvgPrecisionPair value = ObjectSerDeUtils.AVG_PRECISION_PAIR_SER_DE.deserialize(bytesValues[i]);
+          avgPair.apply(value);
+        }
+        break;
+      default:
+        throw new IllegalStateException("Unsupported value type: " + blockValSet.getValueType());
+    }
+    updateAggregationResult(aggregationResultHolder, avgPair);
+  }
+
+  protected void updateAggregationResult(AggregationResultHolder aggregationResultHolder, AvgPrecisionPair avgPair) {
+    if (_nullHandlingEnabled) {
+      if (avgPair != null && avgPair.getCount() > 0) {
+        AvgPrecisionPair otherPair = aggregationResultHolder.getResult();
+        if (otherPair == null) {
+          aggregationResultHolder.setValue(avgPair);
+        } else {
+          otherPair.apply(avgPair);
+        }
+      }
+    } else {
+      if (avgPair == null) {
+        avgPair = new AvgPrecisionPair();
+      }
+      AvgPrecisionPair otherPair = aggregationResultHolder.getResult();
+      if (otherPair == null) {
+        aggregationResultHolder.setValue(avgPair);
+      } else {
+        otherPair.apply(avgPair);
+      }
+    }
```

Comment:
Could you please explain the null handling logic here? You can check the regular average function implementation for pointers.